crates/curp/src/server/raw_curp/log.rs

```diff
@@ -113,19 +120,57 @@ impl<C: Command> LogEntryVecDeque<C> {
     fn push_back(&mut self, entry: Arc<LogEntry<C>>) -> Result<(), bincode::Error> {
         let entry_size = serialized_size(&entry)?;
 
+        if entry_size > self.batch_limit {
+            warn!("entry_size of an entry > batch_limit, which may be too small.",);
+        }
+
         self.entries.push_back(entry);
-        let Some(&pre_entries_size) = self.batch_index.back() else {
-            unreachable!("batch_index cannot be None")
-        };
-        self.batch_index
-            .push_back(pre_entries_size.overflow_add(entry_size));
+        self.entry_size.push_back(entry_size);
+        self.batch_index.push_back(0); // placeholder
+        self.last_batch_size += entry_size;
+
+        while self.last_batch_size >= self.batch_limit {
+            if let Some(cur_batch_index) = self.batch_index.get_mut(self.first_entry_at_last_batch)
+            {
+                *cur_batch_index = self.entries.len() - 1 - self.first_entry_at_last_batch;
+
+                if self.last_batch_size == self.batch_limit {
+                    // self.last_batch_size == self.batch_limit means the last index must be included in `get_range`
+                    *cur_batch_index += 1;
+                }
+
+                if let Some(cur_entry_size) = self.entry_size.get(self.first_entry_at_last_batch) {
+                    self.last_batch_size -= *cur_entry_size;
+                }
+                self.first_entry_at_last_batch += 1;
+            }
+        }
```

Comment:
Please try simpify this procedure, base on previous comment: https://github.com/xline-kv/Xline/pull/704#issuecomment-2034304554

Reply:
It's simpler now